File: iOS/starflower/Shared/WeatherService.swift
//
//  WeatherService.swift
//  starflower
//
//  Created by 양지성 on 6/19/26.
//

import Foundation

// ── 응답 데이터 구조 ──────────────────────────────────────
struct WeatherResponse: Decodable {
    let current: CurrentWeather
    let hourly: HourlyWeather
    let daily: DailyWeather
}

struct CurrentWeather: Decodable {
    let time: String
    let temperature2m: Double
    let relativeHumidity2m: Double
    let cloudCover: Double
    let windSpeed10m: Double
    let surfacePressure: Double
    let weatherCode: Int
    let isDay: Int

    enum CodingKeys: String, CodingKey {
        case time
        case temperature2m        = "temperature_2m"
        case relativeHumidity2m   = "relative_humidity_2m"
        case cloudCover           = "cloud_cover"
        case windSpeed10m         = "wind_speed_10m"
        case surfacePressure      = "surface_pressure"
        case weatherCode          = "weather_code"
        case isDay                = "is_day"
    }
}

struct HourlyWeather: Decodable {
    let time: [String]
    let cloudCover: [Double]
    let relativeHumidity2m: [Double]
    let windSpeed10m: [Double]
    let temperature2m: [Double]
    let weatherCode: [Int]

    enum CodingKeys: String, CodingKey {
        case time
        case cloudCover           = "cloud_cover"
        case relativeHumidity2m   = "relative_humidity_2m"
        case windSpeed10m         = "wind_speed_10m"
        case temperature2m        = "temperature_2m"
        case weatherCode          = "weather_code"
    }
}

struct DailyWeather: Decodable {
    let time: [String]
    let sunrise: [String]
    let sunset: [String]
}

struct AirResponse: Decodable {
    let hourly: AirHourly
}

struct AirHourly: Decodable {
    let time: [String]
    let pm25: [Double?]

    enum CodingKeys: String, CodingKey {
        case time
        case pm25 = "pm2_5"
    }
}

// ── WeatherService ────────────────────────────────────────
final class WeatherService {

    static let shared = WeatherService()
    private init() {}

    private let session = URLSession.shared
    private let decoder = JSONDecoder()

    // 날씨 데이터 fetch
    func fetchWeather(lat: Double, lng: Double) async throws -> WeatherResponse {
        var components = URLComponents(string: "https://api.open-meteo.com/v1/forecast")!
        components.queryItems = [
            .init(name: "latitude",       value: String(lat)),
            .init(name: "longitude",      value: String(lng)),
            .init(name: "current",        value: "temperature_2m,relative_humidity_2m,cloud_cover,wind_speed_10m,surface_pressure,weather_code,is_day"),
            .init(name: "hourly",         value: "cloud_cover,relative_humidity_2m,wind_speed_10m,temperature_2m,weather_code"),
            .init(name: "daily",          value: "sunrise,sunset"),
            .init(name: "wind_speed_unit",value: "ms"),
            .init(name: "timezone",       value: "auto"),
            .init(name: "forecast_days",  value: "4"),
            .init(name: "past_days", value: "1"),
        ]

        let (data, response) = try await session.data(from: components.url!)
        guard (response as? HTTPURLResponse)?.statusCode == 200 else {
            throw ServiceError.httpError
        }
        return try decoder.decode(WeatherResponse.self, from: data)
    }

    // 대기질(PM2.5) fetch — 실패해도 앱 진행 (nil 반환)
    func fetchAir(lat: Double, lng: Double) async -> AirResponse? {
        var components = URLComponents(string: "https://air-quality-api.open-meteo.com/v1/air-quality")!
        components.queryItems = [
            .init(name: "latitude",      value: String(lat)),
            .init(name: "longitude",     value: String(lng)),
            .init(name: "hourly",        value: "pm2_5"),
            .init(name: "timezone",      value: "auto"),
            .init(name: "forecast_days", value: "4"),
        ]

        guard let url = components.url,
              let (data, _) = try? await session.data(from: url),
              let result = try? decoder.decode(AirResponse.self, from: data)
        else { return nil }

        return result
    }
}

// ── 에러 타입 ─────────────────────────────────────────────
enum ServiceError: Error, LocalizedError {
    case httpError
    case noData

    var errorDescription: String? {
        switch self {
        case .httpError: return "날씨 정보를 불러오지 못했어요."
        case .noData:    return "데이터가 없어요."
        }
    }
}

File: Windows_Linux/starflower-desktop/src/main/kotlin/com/songsari/starflower/data/WeatherService.kt
package com.songsari.starflower.data

import com.squareup.moshi.Json
import com.squareup.moshi.Moshi
import com.squareup.moshi.kotlin.reflect.KotlinJsonAdapterFactory
import retrofit2.Retrofit
import retrofit2.converter.moshi.MoshiConverterFactory
import retrofit2.http.GET
import retrofit2.http.Query

// ── 응답 데이터 구조 ──────────────────────────────────────
data class WeatherResponse(
    val current: CurrentWeather,
    val hourly: HourlyWeather,
    val daily: DailyWeather,
)

data class CurrentWeather(
    val time: String,
    @Json(name = "temperature_2m") val temperature2m: Double,
    @Json(name = "relative_humidity_2m") val relativeHumidity2m: Double,
    @Json(name = "cloud_cover") val cloudCover: Double,
    @Json(name = "wind_speed_10m") val windSpeed10m: Double,
    @Json(name = "surface_pressure") val surfacePressure: Double,
    @Json(name = "weather_code") val weatherCode: Int,
    @Json(name = "is_day") val isDay: Int,
)

data class HourlyWeather(
    val time: List<String>,
    @Json(name = "cloud_cover") val cloudCover: List<Double>,
    @Json(name = "relative_humidity_2m") val relativeHumidity2m: List<Double>,
    @Json(name = "wind_speed_10m") val windSpeed10m: List<Double>,
    @Json(name = "temperature_2m") val temperature2m: List<Double>,
    @Json(name = "weather_code") val weatherCode: List<Int>,
)

data class DailyWeather(
    val time: List<String>,
    val sunrise: List<String>,
    val sunset: List<String>,
)

data class AirResponse(val hourly: AirHourly)

data class AirHourly(
    val time: List<String>,
    @Json(name = "pm2_5") val pm25: List<Double?>,
)

// ── Retrofit API ──────────────────────────────────────────
private interface OpenMeteoApi {
    @GET("v1/forecast")
    suspend fun forecast(
        @Query("latitude") lat: Double,
        @Query("longitude") lng: Double,
        @Query("current") current: String,
        @Query("hourly") hourly: String,
        @Query("daily") daily: String,
        @Query("wind_speed_unit") windUnit: String,
        @Query("timezone") timezone: String,
        @Query("forecast_days") forecastDays: Int,
        @Query("past_days") pastDays: Int,
    ): WeatherResponse
}

private interface AirQualityApi {
    @GET("v1/air-quality")
    suspend fun air(
        @Query("latitude") lat: Double,
        @Query("longitude") lng: Double,
        @Query("hourly") hourly: String,
        @Query("timezone") timezone: String,
        @Query("forecast_days") forecastDays: Int,
    ): AirResponse
}

// ── WeatherService ─────────────────────────────────────────
object WeatherService {

    private val moshi = Moshi.Builder()
        .add(KotlinJsonAdapterFactory())
        .build()

    private val weatherApi: OpenMeteoApi = Retrofit.Builder()
        .baseUrl("https://api.open-meteo.com/")
        .addConverterFactory(MoshiConverterFactory.create(moshi))
        .build()
        .create(OpenMeteoApi::class.java)

    private val airApi: AirQualityApi = Retrofit.Builder()
        .baseUrl("https://air-quality-api.open-meteo.com/")
        .addConverterFactory(MoshiConverterFactory.create(moshi))
        .build()
        .create(AirQualityApi::class.java)

    suspend fun fetchWeather(lat: Double, lng: Double): WeatherResponse =
        weatherApi.forecast(
            lat = lat,
            lng = lng,
            current = "temperature_2m,relative_humidity_2m,cloud_cover,wind_speed_10m,surface_pressure,weather_code,is_day",
            hourly = "cloud_cover,relative_humidity_2m,wind_speed_10m,temperature_2m,weather_code",
            daily = "sunrise,sunset",
            windUnit = "ms",
            timezone = "auto",
            forecastDays = 4,
            pastDays = 1,
        )

    /** 대기질은 실패해도 진행 (null 반환) */
    suspend fun fetchAir(lat: Double, lng: Double): AirResponse? =
        try {
            airApi.air(
                lat = lat,
                lng = lng,
                hourly = "pm2_5",
                timezone = "auto",
                forecastDays = 4,
            )
        } catch (e: Exception) {
            null
        }
}
